### PR TITLE
document StaticIconProvider

### LINKED ISSUE #120459: Document the @staticIconProvider annotation in the `IconData` class dartdoc
Users (usually package authors) should know about this annotation, which will help icon tree-shaking on web.

### PR DESCRIPTION
**Pull request by christopherfujino**
Fixes https://github.com/flutter/flutter/issues/120459

### CODE REVIEW

**Comment by christopherfujino** [comment]
> (triage) @christopherfujino Do you still have plans to finish this one up? Would be awesome to have better documentation on this!

Yup, thanks for the ping!

**Comment by christopherfujino** [comment]
ping @christopherfujino please land this.

**Comment by christopherfujino** [comment]
@goderbauer this is my first framework doc PR, does this align with our style?

File: packages/flutter/lib/src/widgets/icon_data.dart [hunk 1]
```diff
@@ -17,6 +21,9 @@ class IconData {
   ///
   /// The [fontPackage] argument must be non-null when using a font family that
   /// is included in a package. This is used when selecting the font.
+  ///
+  /// Instantiating non-const instances of this class can cause icon
+  /// tree-shaking to fail.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/icon_data.dart:26]: Link this to something where people can learn more about icon tree shaking?

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/icon_data.dart:26]: Also, maybe document what failure means here: Will your app stop running/building? Is it that your app will function just fine, but the app size will be larger than necessary?

File: packages/flutter/lib/src/widgets/icon_data.dart [hunk 2]
```diff
@@ -99,6 +106,18 @@ class _StaticIconProvider {
 /// Annotation for classes that only provide static const [IconData] instances.
 ///
 /// This is a hint to the font tree shaker to ignore the constant instances
-/// of [IconData] appearing in the class when tracking which code points
-/// should be retained in the bundled font.
+/// of [IconData] appearing in the declaration of this class when tree-shaking
+/// unused code points from the bundled font.
+///
+/// Classes with this annotation should only have static const declarations of
+/// the [IconData] class, such as:
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/icon_data.dart:113]: nit: I would rephrase this, instead of saying "should" I would say, tree shaking only works if you do this. Otherwise your binary will be larger than necessary (or whatever the failure mode is).

File: packages/flutter/lib/src/widgets/icon_data.dart (hunk 1 above)
Comment (christopherfujino) [reply packages/flutter/lib/src/widgets/icon_data.dart:26]: ack about documenting the failiure mode.

AFAIK this is the first documentation we have about the icon tree-shaking feature. Should I link to the source?

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/icon_data.dart:26]: Maybe link it to staticIconProvider? That seems to be the canonical doc for what this tree shaking thing is for icons?

Comment (christopherfujino) [reply packages/flutter/lib/src/widgets/icon_data.dart:26]: Good idea.

File: packages/flutter/lib/src/widgets/icon_data.dart (hunk 2 above)
Comment (christopherfujino) [reply packages/flutter/lib/src/widgets/icon_data.dart]: updated, WDYT?

File: packages/flutter/lib/src/widgets/icon_data.dart (hunk 1 above)
Comment (christopherfujino) [reply packages/flutter/lib/src/widgets/icon_data.dart]: Done and done.

File: packages/flutter/lib/src/widgets/icon_data.dart [hunk 3]
```diff
@@ -17,6 +21,11 @@ class IconData {
   ///
   /// The [fontPackage] argument must be non-null when using a font family that
   /// is included in a package. This is used when selecting the font.
+  ///
+  /// Instantiating non-const instances of this class will in your app will
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/icon_data.dart:25]: nit: remove one of the `will`s?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/icon_data.dart (hunk 3 above)
Comment (christopherfujino) [inline packages/flutter/lib/src/widgets/icon_data.dart:25]:
```suggestion
  /// Instantiating non-const instances of this class in your app will
```

Comment (christopherfujino) [reply packages/flutter/lib/src/widgets/icon_data.dart]: Good catch

**Comment by slavap** [comment]
Reading comment added for StaticIconProvider still does not answer the question - "How to make tree shaking working for custom font?". I have custom icon font added to assests and declared:
```dart
@staticIconProvider
abstract final class MyIcons { ... }
```
but don't see tree shaking works for my custom font. What should I do to make it work?
